Focal loss

arcunliffe opened the pull request.

Commit:
first pass at implementing focal loss

Commit:
make it work for binary classification

Commit:
now it works for multi-label

Commit:
refactor to expose do_focal_loss and focal_loss_gamma as parameters in train_multilabel_classifier

Commit:
add new options to train.py example

Commit:
remove unused set_class_weights

Commit:
remove call to set_class_weights

Commit:
remove set_class_weights from inference_utils.py

Commit:
update focal loss to use torch BCE

Commit:
remove unused constant USE_FOCAL_LOSS

Commit:
clean up

Commit:
switch order so that focal loss gamma is set correctly in focal loss

Commit:
remove unused imports

jmpounders commented:
This looks good to me.  Want to go ahead a merge?  I will add the new loss stuff in after.

Commit:
Merge branch 'main' into focal-loss

Commit:
remove block_size from training_arguments to avoid error

Newer version of transformers checks to make sure all training_arguments are valid when parsing, so we can't pass block_size here anymore

Commit:
remove accidental unused import

arcunliffe commented:
> This looks good to me. Want to go ahead a merge? I will add the new loss stuff in after.

OK great! I haven't made changes to this repo in a while, and the CI workflow failed due to some newer changes both within GitHub (I think) and transformers. Specifically:

- The python 3.6 workflow doesn't seem to work anymore in GitHub. I didn't look too much into this but instead removed support for python 3.6 since I don't think we need that anymore?
- Newer transformers versions check the training_arguments when parsing them to make sure they are all valid. I was previously including "block_size" in the dictionary of training_arguments and using it elsewhere in the workflow. It didn't matter that it wasn't an expected training argument for the HF Trainer, since it was just ignored there. Now that fails, so I've added DATASET_ARGUMENTS and specified "block_size" as a key in this dictionary instead.

Take one more look and let me know if you're ok to merge with these additional changes. I can merge, or you can if you think it all looks good!
